### PR TITLE
Record a metric for duration and rate of execution for every flow

### PR DESCRIPTION
**Pull request by adagys**
I intended to use this for performance tests only, but think it will be a useful metric for test/production deployments to detect performance degradation.

This exports a metric via JMX with the rate and duration distribution, which we can graph and potentially alert on.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt [hunk 1]
```diff
@@ -381,4 +384,16 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
...
         }
     }
+
+    /**
+     * Records the duration of this flow – from call() to completion or failure.
+     * Note that the duration will include the time the flow spent being parked, and not just the total
+     * execution time.
+     */
+    private fun recordDuration(startTime: Long) {
+        val timer = serviceHub.monitoringService.metrics.timer("FlowDuration.${logic.javaClass}")
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: I'm assuming this just gives the class name (fully qualified)?  Is it perhaps better to explicitly use `logic.javaClass.name` ?

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt [hunk 2]
```diff
@@ -102,6 +104,8 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
             logger.warn("Terminated by unexpected exception", t)
             processException(t, false)
             return
+        } finally {
+            recordDuration(startTime)
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: Should we distinguish between success and failure?  The timing could be very different.  Could have `FlowDuration.Success.<class>` and `FlowDuration.Failure.<class>`.  Would that also allow us to see numbers of successful flows vs. unsuccessful too ?

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt [hunk 3]
```diff
@@ -381,4 +384,16 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
...
+    /**
+     * Records the duration of this flow – from call() to completion or failure.
+     * Note that the duration will include the time the flow spent being parked, and not just the total
+     * execution time.
+     */
+    private fun recordDuration(startTime: Long) {
+        val timer = serviceHub.monitoringService.metrics.timer("FlowDuration.${logic.javaClass}")
+        // Start time gets serialized along with the fiber when it suspends
+        val duration = System.nanoTime() - startTime
+        timer.update(duration, TimeUnit.NANOSECONDS)
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: Does this give us rate as well as duration as the subject suggests?

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt (hunk 2 above)
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: Yes sounds good

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt (hunk 3 above)
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: Yeah, it gives the min/max/avg duration, percentiles, and 1/5/15 min rates

**Comment by adagys** [comment]
@rick-r3 back to you

**Pull request merged**